text/3351-relax-const-restrictions.md

```diff
@@ -0,0 +1,99 @@
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Pure `const fn` under the old rules can be seen as a simple optimization opportunity for naive optimizers, as they could just reuse constant evaluation for `const fn` if the argument is known at compile time, even if the function is in a runtime context. This RFC makes such an optimization impossible. This is not seen as a problem by the author, as a more advanced optimizer (like LLVM) is able to remove these calls at compile time through means other than Rust's constant evaluation (inlining and constant folding). Also, a constant evaluation system can still evaluate executions in a runtime context, as long as it behaves exactly like runtime. The optimizer could also manually annotate functions as being truly pure by looking at the body.
+
+Secondly, with the current rules around `const fn` purity, unsafe code could choose to rely on purity, e.g. by caching function return values and assuming this is not observable to clients. The author does not see this as a significant drawback, as this functionality is better served by language features that target this use case directly (like a `pure` attribute) and is therefore out of scope for the language feature of "functions evaluatable at compile time". This could break code that already relies on this, but since Rust doesn't have proper support for this, the impact should be minimal at most.
+
+The old rules, which say that `const fn` always has to behave the same way are already well-known in the community. Changing this will require teaching people about the new change. Since this is a simple change, this should not be too hard (for example with a mention in the release notes).
+
+This is technically a breaking change. Code could rely on this behavior right now, as the [internal documentation](https://doc.rust-lang.org/1.65.0/std/intrinsics/fn.const_eval_select.html#safety) for `std::intrinsics::const_eval_select` explains. Relying on this was never endorsed or officially documented and there are no known cases of code relying on it. This is deemed to be highly unlikely and even if some code did rely on this, it will continue to work as long as no new behavioral differences are introduced by the code. The internal docs will have to be adjusted after this RFC is accepted.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+An alternative is to keep the current rules. This is bad because the current rules are highly restrictive and don't have a significant benefit to them. With the current rules, floats cannot be used in `const fn` without significant restrictions.
+
+It would also be possible to allow them to behave differently, but keep the restrictions around purity and determinism at runtime. This would still allow unsafe code to treat `const fn` specially, but this is not seen as a desirable feature of `const fn`.
+
```

Comment:
```suggestion

An alternative would be to implement floats for compile-time evaluation in a manner always identical to the target's floating-point behavior.

```

(As well as added text for how feasible that would be. It doesn't seem impossible, though it certainly seems difficult. But, for instance, as far as I know instruction emulators do this successfully.)

Reply:
It's impossible on some targets such as wasm that do not have deterministic floating point behavior. There also might be issues with targets where different CPUs for the same target have different behavior (though I do not know if such targets/CPUs actually exist).

Reply:
It's easier than it sounds, since IEEE 754-compliant targets may only differ in which NaNs they produce. But yes, LLVM is currently very noncompliant, between failing to preserve NaN bits under copies and just having value-changing transformations.

Reply:
On wasm, [quoting the spec](https://webassembly.github.io/spec/core/exec/numerics.html#floating-point-operations):

> Floating-point arithmetic follows the IEEE 754-2019 standard, with the following qualifications:
>
> - All operators use round-to-nearest ties-to-even, except where otherwise specified. Non-default directed rounding attributes are not supported.
> - Following the recommendation that operators propagate NaN payloads from their operands is permitted but not required.
> - All operators use “non-stop” mode, and floating-point exceptions are not otherwise observable. In particular, neither alternate floating-point exception handling attributes nor operators on status flags are supported. There is no observable difference between quiet and signalling NaNs.

This specification places no restrictions on what NaNs are produced by NaN-producing operations, because the IEEE 754-2019 standard does not. It is thus impossible for rustc to provide bit-equivalent float operations to the target, because the target does not have a single definition to match.

As for LLVM: the float optimizations which rustc enables are IIRC all IEEE-precise, assuming that the default floating point environment is always used and never changed. We do not enable any `-ffast-math`-style optimization which is allowed to reorder, contract, expand, or otherwise transform floating point operations in a way that impacts the precision of the result.

LLVM is not, however, required to produce the target-accurate NaN bits when constant folding, as IEEE places no restrictions on what precise NaN is produced by any NaN-producing value.

The only nondeterminism in the IEEE standard is the behavior of NaNs, but because Rust allows inspecting the bit value of NaNs, this is not sufficient for `const` unless *perhaps* if you want to treat bitcast-float-to-int like ptr-to-int.

The sensical behavior for Rust `const` evaluation is probably to force all floating point NaNs to be a single canonical NaN. (For wasm, the canonical NaN significand is all bits clear except the MSB (the quiet bit). wasm IIUC considers both positive and negative NaNs canonical; RISC-V on the other hand names only the positive canonical quiet NaN as being canonical.)

Reply:
Suggested wording:

```suggestion

An alternative could be to implement floats for compile-time evaluation in a manner always identical to the target's floating-point behavior. Unfortunately, some targets such as wasm do not have fully deterministic floating point behavior to be emulated, so while rustc could theoretically guarantee that compile-time float evaluation matches *an* implementation of the target, it would still have to choose one of multiple potential implementations of the target to emulate. Wasm is a convenient example here, as the wasm specification only requires floating point operations to be accurate to the IEEE standard and places essentially no further restrictions, leaving the choice of specific NaN production rules up to the implementor.

```

Reply:
There is, in the very least, one invalid floating-point optimization that `rustc` performs: https://godbolt.org/z/5qa5Go17M. It transforms `x * 1.0` into just `x`, which is invalid whenever `x` is a signaling NaN. IEEE 754 mandates that the result of that operation is a quiet NaN.

I do advocate for abstract (or symbolic, or whatever paradigm you prefer) interpretation of nondeterministic code, but I know it's harder if the interpreter wasn't built for it from the ground up. Ultimately, const evaluation of floating-point expressions should always give the same result everywhere. We can very easily achieve that sort of consistency now by erroring out whenever an expression evaluates to a nondeterministic NaN. That would cover almost all use cases, and would not be a severe restriction on compile-time floating-point code. This is why I don't believe that const-eval of floating-point is a good motivation per se for this RFC.

Reply:
> There is, in the very least, one invalid floating-point optimization that `rustc` performs: [godbolt.org/z/5qa5Go17M](https://godbolt.org/z/5qa5Go17M). It transforms `x * 1.0` into just `x`, which is invalid whenever `x` is a signaling NaN. IEEE 754 mandates that the result of that operation is a quiet NaN.

Can you file a bug for this? I'm not sure it will be fixed with urgency, but it's worth having on file.

Reply:
Filed under https://github.com/rust-lang/rust/issues/107247. I'm not holding my breath for it getting fixed any time soon though, it's really an LLVM bug.

Reply:
Interestingly enough, MSVC [does the same thing](https://cpp.godbolt.org/z/j8Mhshncc) even with [`/fp:strict`](https://learn.microsoft.com/en-us/cpp/build/reference/fp-specify-floating-point-behavior?view=msvc-170). This makes me less inclined to believe this is a miscompilation.

I've access to the actual text of IEEE 754-2019 through my current university (I suppose I could've grabbed 754-2008 instead, but I grabbed the latest revision), so here's the actual relevant quotes:

> §6.2 ¶3: Signaling NaNs shall be reserved operands that signal the invalid operation exception (see §7.2) for every general-computational and signaling-computational operation except for the conversions described in §5.12.

> §6.2 ¶2: Under default exception handling, any operation signaling an invalid operation exception and for which a floating-point result is to be delivered, except as stated otherwise, shall deliver a quiet NaN. For non-default treatment, see §8.

> §7.2 ¶2: For operations producing results in floating-point format, the default result of an operation that signals the invalid operation exception shall be a quiet NaN that should provide some diagnostic information (see §6.2). These operations are:
> a) any general-computational operation on a signaling NaN (see §6.2), except for some conversions (see §5.12)

§5.12 is about conversions between floating-point data and character sequences, so is not relevant. Multiplication is a "*formatOf* general-computational operation".

Thus, @Muon: I believe you're correct in asserting that the standard mandates that `signalingNaN * 1.0` should (raise an invalid operation exception and) return a quiet NaN, and that the transformation to just return `signalingNaN` is incorrect.

If we relax the specification from "default floating point environment" to "implementation defined floating point environment" *and* allow the language to arbitrarily change the exception handling mode, then we can justify the transformation, by first setting the exception handling mode for invalid operation to `substitute(signalingNaN)`, doing `signalingNaN * 1.0` (signaling an invalid operation exception, handling that by substituting `signalingNaN` as the result), and then setting the exception handling mode back to whatever it was previously (probably `mayRaiseFlag` in this world).

FWIW, alive2 [times out for `float`](https://alive2.llvm.org/ce/z/fyPmj_) and [considers the transformation valid for `half`](https://alive2.llvm.org/ce/z/z7f5bm). [From LLVM LangRef](https://llvm.org/docs/LangRef.html#floating-point-environment):

> The default LLVM floating-point environment assumes that floating-point instructions do not have side effects. Results assume the round-to-nearest rounding mode. No floating-point exception state is maintained in this environment. Therefore, there is *no attempt to* create or *preserve invalid operation (SNaN)* or division-by-zero exceptions [emphasis mine].
>
> The benefit of this exception-free assumption is that floating-point operations may be speculated freely without any other fast-math relaxations to the floating-point model.

I [tried using](https://alive2.llvm.org/ce/z/KKGXPm) the (experimental?) [constrained floating point intrinsics](https://llvm.org/docs/LangRef.html#constrained-floating-point-intrinsics), but I don't think alive2 supports the `strictfp` attribute. As far as I can tell, we need `!"fpexcept.strict"` in order to avoid removing operations which may raise a floating point exception. But even with `!"fpexcept.ignore"`, that's just defined as "may assume that the exception status flags will not be read and that floating-point exceptions will be masked", which I don't think is strong enough to assume signaling NaNs don't exist (invalid operation exceptions won't be raised), and the change from signaling to quiet should need to be preserved under those semantics.

Of note from the Rust docs, though, is that a bunch of `f32` methods say roughly

> This follows the IEEE 754-2008 semantics for *operation*, except for handling of signaling NaNs; this function handles all NaNs the same way [...]

I think there's a fair chance that the Rust opsem will make no distinction between signaling and quiet NaNs. I can somewhat justify a perverse reading of IEEE 754-1985 that allows all binary encoded NaNs to be quiet; 754-2008 and 754-2019 of course mandate a specific interpretation of the mantissa MSB as indicating quiet (set) or signaling (cleared), but Rust does support targets such as MIPS which do not agree with the 2008 definition.

Reply:
That's missing the point. Const eval could do any number of things to allow generating NaNs. But if you allow it without restrictions, you can trivially get a const item whose value depends on the compilation target, without using any conditional compilation tricks.

Reply:
@CAD97 I like that wording, though it should also point out that LLVM makes no attempt at guaranteeing deterministic NaN behavior even if the target has deterministic NaNs, so even for non-wasm targets making such a guarantee is currently not practical for Rust.

Reply:
I'd then add another sentence

> Similar goes for LLVM and most other code optimization backends, which only aim to preserve the abstract IEEE semantics and not the actual target's behavior over transformations, and may thus change the computed NaN in an unpredictable manner.

(although I'm not a big fan of the way this sentence starts.)